File: clients/windows/YunQiao.Cashier/src/YunQiao.Cashier.Core/Protocol/ServerPayloadIntegrity.cs
using System.Security.Cryptography;

namespace YunQiao.Cashier.Core.Protocol;

public static class ServerPayloadIntegrity
{
    public static byte[] DecodeBase64(string encoded, int declaredLength, string? expectedSha256)
    {
        byte[] payload;
        try { payload = Convert.FromBase64String(encoded); }
        catch (FormatException error)
        {
            throw new TerminalApiException(200, "PAYLOAD_INTEGRITY_FAIL", "Server payload base64 is invalid.", error);
        }
        var actual = Convert.ToHexString(SHA256.HashData(payload)).ToLowerInvariant();
        if (payload.Length != declaredLength || expectedSha256?.Length != 64 ||
            !string.Equals(actual, expectedSha256, StringComparison.Ordinal))
            throw new TerminalApiException(200, "PAYLOAD_INTEGRITY_FAIL", "Server payload hash mismatch.");
        return payload;
    }
}

File: apps/merchant-terminal-v2-android/app/src/main/java/com/yunqiao/life/merchantterminal/network/ServerPayloadIntegrity.kt
package com.yunqiao.life.merchantterminal.network

import java.security.MessageDigest
import java.util.Base64

object ServerPayloadIntegrity {
    private val SHA256 = Regex("^[a-f0-9]{64}$")

    fun decodeBase64(encoded: String, declaredLength: Int, expectedSha256: String?): ByteArray {
        val payload = try {
            Base64.getDecoder().decode(encoded)
        } catch (_: IllegalArgumentException) {
            throw V2ApiException(200, "PAYLOAD_INTEGRITY_FAIL", message = "Server payload base64 is invalid.")
        }
        val actual = MessageDigest.getInstance("SHA-256")
            .digest(payload)
            .joinToString("") { "%02x".format(it) }
        if (
            payload.size != declaredLength ||
            expectedSha256 == null ||
            !SHA256.matches(expectedSha256) ||
            actual != expectedSha256
        ) {
            throw V2ApiException(200, "PAYLOAD_INTEGRITY_FAIL", message = "Server payload hash mismatch.")
        }
        return payload
    }
}
